### PR TITLE
feat(#64) added multiple aggregate levels

### PR DESCRIPTION
**Pull request by mscottx88**
@jasonwilson Here is the first go at this code.  I need to add unit tests for the new work.  If you get a moment, check it out and see what you think!

### CODE REVIEW

File: lib/providers/metrics-provider.js [hunk 1]
```diff
@@ -1,25 +1,461 @@
...
+    this.minimumAggregationInterval = +this.aggregationLevels[0];
+
+    this._startTime = Date.now();
+    this._lastAggregation = Date.now();
+
+    screen.on("metrics", this._onMetrics.bind(this));
+
+    screen.on("zoomAggregate", function zoomAggregate(zoom) {
+      // apply zoom delta and check for boundaries
+      this.currentAggregateZoom += zoom;
```
Comment (jasonwilson) [inline lib/providers/metrics-provider.js]: Could we just use something like lodash.clamp (https://lodash.com/docs/4.17.4#clamp) to enforce bounds?

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{
+  // Use IntelliSense to learn about possible Node.js debug attributes.
```
Comment (aisapatino) [inline .vscode/launch.json]: i think this is a visual studio file - should add file/dir to `.gitignore`

File: lib/dashboard.js [hunk 3]
```diff
@@ -94,8 +100,8 @@ var VIEW_MAP = {
   eventLoop: EventLoopView
 };
 
-Dashboard.prototype._showLayout = function (id) {
-  if (this.currentLayout === id) {
+Dashboard.prototype._showLayout = function (id, forced) {
```
Comment (mscottx88) [inline lib/dashboard.js]: I'm sure there is.  I will work on it!

File: lib/providers/metrics-provider.js (hunk 1 above)
Comment (mscottx88) [inline lib/providers/metrics-provider.js]: 👍

File: lib/providers/metrics-provider.js [hunk 4]
```diff
@@ -1,25 +1,461 @@
 "use strict";
 
 var EventEmitter = require("events").EventEmitter;
+var config = require("../config.js");
 
-var MetricsProvider = function MetricsProvider(screen) {
-  EventEmitter.call(this);
+// time scaling factors
+var timeScales = [
```
Comment (mscottx88) [inline lib/providers/metrics-provider.js]: 👍

File: gulpfile.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+"use strict";
```
Comment (ryan-roemer) [inline gulpfile.js]: Small nit here -- at least as a broad Formidable convention (and same for most of our clients), npm scripts have "won".

Gulp, while a great tool, has persistent problems of plugins being behind the "real" upstream projects and adding complexity / abstraction where it's not needed. Here, we just want to execute mocha. I vote we stick with mocha + `package.json:scripts` until we hit a use case we truly can't support.

Side note -- if we need concurrent execution of things, we've got great tools for that that work with `package.json:scripts`.

@jasonwilson -- Thoughts on ^^^ ?

File: test/lib/dashboard-agent.spec.js [hunk 6]
```diff
@@ -64,18 +66,38 @@ describe("dashboard-agent", function () {
 
       var checkMetrics = function (metrics) {
         expect(metrics).to.be.an("object");
-        expect(metrics.eventLoop).to.deep.equal({ delay: 0, high: 0 });
-        expect(metrics.mem.systemTotal).to.be.above(0);
-        expect(metrics.mem.rss).to.be.above(0);
-        expect(metrics.mem.heapTotal).to.be.above(0);
-        expect(metrics.mem.heapUsed).to.be.above(0);
-        expect(metrics.cpu.utilization).to.be.above(0);
+        expect(metrics.eventLoop.delay).to.be.a("number");
```
Comment (ryan-roemer) [inline test/lib/dashboard-agent.spec.js]: GLOBAL SUGGESTION: I know this is a refactor, but for things like:

```js
expect(metrics.eventLoop.delay).to.be.a("number");
```

if for example, `metrics.eventLoop` is `undefined`, you're going to get a not-so-useful assertion failure. A better way to do this is to assert on the property and let chai tell you "that deep property doesn't exist". Aka

```js
expect(metrics).to.have.deep.property("eventLoop.delay").that.is.a("number");
```

(Side note -- unique to `.property` assertions, the context changes from there on to the _property_ asserted against (e.g. `metrics.eventLoop.delay`) and not the original target object (e.g. `metrics`))

File: gulpfile.js (hunk 5 above)
Comment (mscottx88) [inline gulpfile.js]: Can npm run scripts allow you to debug your spec files? If so, that is great. Otherwise, the only purpose for the gulp is for debugging unit tests and I can keep it out of the main repo.

Although, excluding it would make portability an issue across dev machines.

File: test/lib/dashboard-agent.spec.js [hunk 7]
```diff
@@ -64,18 +66,38 @@ describe("dashboard-agent", function () {
...
+        expect(metrics.eventLoop.high).to.be.a("number");
+        expect(metrics.mem.systemTotal).to.equal(20);
+        expect(metrics.mem.rss).to.equal(30);
+        expect(metrics.mem.heapTotal).to.equal(40);
+        expect(metrics.mem.heapUsed).to.equal(50);
+        expect(metrics.cpu.utilization).to.equal(60);
       };
 
+      var stubProcess = sinon.stub(process, "memoryUsage", function () {
```
Comment (ryan-roemer) [inline test/lib/dashboard-agent.spec.js]: GLOBAL: Never use `sinon.stub|spy` on anything that requires a `.restore()`. Instead, use a `sandbox` which we declare at top scope for this test suite. 

Here you've got a problem that if anything fails / errors in this test, you won't reach the `.restore()`s on L99-L100. That means you have real methods potentially left unrestored that could cause unrelated tests to fail. Thus, you must always try/catch synchronous stubs to restore _or_ do it in a `beforeEach|afterEach` pair.

Since we already do this with a `sandbox`, you can just _use_ that to `sandbox.stub` and never worry about the restore, since a global restore is done in the top-level `afterEach`

File: test/lib/generate-layouts.spec.js [hunk 8]
```diff
@@ -149,6 +154,8 @@ describe("generate-layouts", function () {
         ]
       }
     ]]);
+
+    done();
```
Comment (ryan-roemer) [inline test/lib/generate-layouts.spec.js]: This looks unnecessary.

File: test/lib/dashboard-agent.spec.js (hunk 6 above)
Comment (mscottx88) [inline test/lib/dashboard-agent.spec.js]: Neat, I like it.

File: test/lib/providers/metrics-provider.spec.js [hunk 9]
```diff
@@ -1,36 +1,242 @@
...
+          valueA: Math.random() * 100
+        },
+        metricB: {
+          valueA: Math.random() * 100,
+          valueB: Math.random() * 100
+        }
+      });
+    }
+
+    done();
```
Comment (ryan-roemer) [inline test/lib/providers/metrics-provider.spec.js]: Same here.

File: test/lib/providers/metrics-provider.spec.js [hunk 10]
```diff
@@ -1,36 +1,242 @@
...
   });
 
-  beforeEach(function () {
+  after(function (done) {
+    done();
```
Comment (ryan-roemer) [inline test/lib/providers/metrics-provider.spec.js]: What is this?

File: test/lib/providers/metrics-provider.spec.js (hunk 9 above)
Comment (ryan-roemer) [inline test/lib/providers/metrics-provider.spec.js]: GLOBAL: `done()` should only be used for async callbacks. Everywhere that it's been added seems synchronous.

File: test/lib/dashboard-agent.spec.js (hunk 7 above)
Comment (mscottx88) [inline test/lib/dashboard-agent.spec.js]: I tried the sandbox but when the sandbox.restore() ran, the stub was still there. On other words, the sandbox did not work as expected.

File: test/lib/views/eventloop-view.spec.js [hunk 11]
```diff
@@ -1,7 +1,10 @@
 "use strict";
 
-var expect = require("chai").expect;
+var chai = require("chai");
 var sinon = require("sinon");
+var sinonChai = require("sinon-chai");
+var expect = chai.expect;
+chai.use(sinonChai);
```
Comment (ryan-roemer) [inline test/lib/views/eventloop-view.spec.js]: GLOBAL: This shouldn't be necessary -- this boilerplate is already taken care of in `setup.js` https://github.com/FormidableLabs/nodejs-dashboard/blob/master/test/setup.js

I think maybe the gulp switch isn't using setup? At any rate go with how we already ran mocha from package.json scripts and should work fine.

File: test/lib/providers/metrics-provider.spec.js (hunk 9 above)
Comment (mscottx88) [inline test/lib/providers/metrics-provider.spec.js]: Sounds good. I've always done the done() as a consistency thing, but easily removed!

File: gulpfile.js (hunk 5 above)
Comment (ryan-roemer) [inline gulpfile.js]: First, you can always use `mocha` directly. Second npm scripts allow extra commands to punch to the command via `--`. So, for example, 

```sh
$ npm run test-only -- --grep=foo --ANY_OTHER_MOCHA_CLI_FLAG
```

works just fine.

Also, we write our npm package scripts to be the portable subset of shell commands that is universally (mac, win, linux) compatible.

File: test/lib/views/eventloop-view.spec.js (hunk 11 above)
Comment (mscottx88) [inline test/lib/views/eventloop-view.spec.js]: I tried npm run scripts in addition to gulp and the problem was still there.

File: test/lib/dashboard-agent.spec.js (hunk 7 above)
Comment (ryan-roemer) [inline test/lib/dashboard-agent.spec.js]: We can work together on Monday. A sandbox restores anything created with it.

Comment (ryan-roemer) [inline test/lib/dashboard-agent.spec.js]: Also, @jasonwilson and @aisapatino can help with this as they've both got a ton of experience with mocha + sandboxes.

Comment (mscottx88) [inline test/lib/dashboard-agent.spec.js]: Aha, I know what I did wrong, instead of sandbox.stub() I used sinon.stub() and that would explain why it didn't go in the sandbox.

I'll correct this one, thanks!

Comment (ryan-roemer) [inline test/lib/dashboard-agent.spec.js]: One thing I did notice was the sandbox create was in a `before()` (pre-existing code). This may be problematic, and as a general rule for unit tests, `beforeEach()` is almost always better than a `before()`. `before()` is typically reserved for super-expensive time-wise things, and I don't see that here. (Same for `after()`)

File: test/lib/views/eventloop-view.spec.js (hunk 11 above)
Comment (ryan-roemer) [inline test/lib/views/eventloop-view.spec.js]: We can work through it on Monday. If we've got you Slack access, just hit me or @jasonwilson up and we'll work it out.

And generally, speaking, anytime you're doing a repeated, fundamental change to a bunch of files, feel free to bounce that off of us early so we can possibly help debug things up front and save you some time!

File: gulpfile.js (hunk 5 above)
Comment (mscottx88) [inline gulpfile.js]: By dev machine, I really meant for those folks that might use VS code as their debugger, these tagalong files are helpful to include

Comment (ryan-roemer) [inline gulpfile.js]: I'm not familiar with VS code's debugging, but I'd be surprised if generic gulp was supported and raw mocha wasn't. In any case, let's keep the main repo gulp-free to save on dependencies we don't strictly need 😉

Comment (mscottx88) [inline gulpfile.js]: Sounds good. Also sounds like I might have another interested party in learning debugging and using VS code?

I think you'll find it to be a really amazing editor.

Comment (mscottx88) [inline gulpfile.js]: Since it is an Electron app, it is built on Node too, so this editor is perfectly adapted to Node development.

Comment (mscottx88) [inline gulpfile.js]: Out of curiosity, since you're not using VS code, what editor are you using?

File: test/lib/views/eventloop-view.spec.js (hunk 11 above)
Comment (mscottx88) [inline test/lib/views/eventloop-view.spec.js]: Hmm I wonder, and this could be a red herring, but is it possibly related to the order of the parameters on the npm script? I see a difference between the coverage script and test-only script.

Comment (mscottx88) [inline test/lib/views/eventloop-view.spec.js]: Scratch that, I see the commands aren't the same.

File: gulpfile.js (hunk 5 above)
Comment (mscottx88) [inline gulpfile.js]: I believe I have the solution I was looking for to remove the gulp and still get the debug ability.

https://gist.github.com/paambaati/54d33e409b4f7cf059cc

Up to date even!

Comment (ryan-roemer) [inline gulpfile.js]: I use many, currently vs code. What I meant is that I don't use the integrated debugger -- typically when I need a debugger I just do the normal node flags and open up chrome.

Nice to see you found what you need and your preferred debugging solution is still available to you!

**Comment by mscottx88** [comment]
@ryan-roemer @jasonwilson @aisapatino I believe this is at last ready for final review :)

**Comment by ryan-roemer** [comment]
@mscottx88 -- Great! I'll review for formalities / test stuff and @jasonwilson @aisapatino can do the substantive stuff in addition to that.

File: package.json [hunk 12]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "nodejs-dashboard",
-  "version": "0.4.1",
+  "version": "0.5.0",
```
Comment (ryan-roemer) [inline package.json]: Again, typically we don't bump versions in PRs. Also, this differs from changelog entry.

File: package.json [hunk 13]
```diff
@@ -16,7 +16,7 @@
     "test": "npm run lint && npm run test-only",
     "test-only": "mocha -c --require test/setup.js --recursive ./test",
     "test-app": "./bin/nodejs-dashboard.js node test/app/index.js",
```
Comment (ryan-roemer) [inline package.json]: Not part of this PR, but can you remove the prefix `./` in this command? `./` is unnecessary on mac/linux and breaks windows.

File: test/lib/generate-layouts.spec.js [hunk 14]
```diff
@@ -5,8 +5,13 @@ var expect = require("chai").expect;
...
 var mockRequire = require("mock-require");
 
+// to ensure cross-platform consistency with mixed Posix & Win32 paths, use normalize()
+// ideally, this would be taken care of as a fix for mockRequire.
+// see https://github.com/boblauer/mock-require/issues/20
+var normalize = require("path").normalize;
+
 var mock = function (path, obj) {
-  return mockRequire(process.cwd() + "/" + path, obj);
+  return mockRequire(normalize(process.cwd() + "/" + path), obj);
```
Comment (ryan-roemer) [inline test/lib/generate-layouts.spec.js]: The existing string concat with `"/"` feels wrong.

Wouldn't 

```js
return mockRequire(path.resolve(path), obj);
```

solve all the problems we have here?

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,5 +1,8 @@
 # Change log
 
+## [v0.5.1] - 2017-07-24
```
Comment (mscottx88) [inline CHANGELOG.md]: yea i meant 5.0.0... sorry about that.  Should I put Unreleased for now or leave CHANGLOG unchanged as part of this PR?

File: package.json (hunk 13 above)
Comment (mscottx88) [inline package.json]: This one is weird.  As is, with the `./`, I get the following error:

> ./bin/nodejs-dashboard.js node test/app/index.js

'.' is not recognized as an internal or external command,
operable program or batch file.

When I change it to not have the `./` I get this error instead:

> nodejs-dashboard@0.4.1 test-app C:\Users\formidable\Documents\nodejs-dashboard
> bin/nodejs-dashboard.js node test/app/index.js

'bin' is not recognized as an internal or external command,
operable program or batch file.

Comment (mscottx88) [inline package.json]: Maybe my npm settings are incorrect.  When I put `node` in front, it works:

$ npm run test-app

> nodejs-dashboard@0.4.1 test-app C:\Users\formidable\Documents\nodejs-dashboard
> node bin/nodejs-dashboard.js node test/app/index.js

Waiting for client connection on 9838...

Comment (mscottx88) [inline package.json]: @ryan-roemer Should/Could it be done as:

"test-app": "node bin/nodejs-dashboard.js -- node test/app/index.js",

Comment (ryan-roemer) [inline package.json]: @jasonwilson -- Can you comment here? The problem is that `bin/nodejs-dashboard.js` isn't a recognizable Windows cmd script.

I'm guessing we want:

```sh
node bin/nodejs-dashboard.js node test/app/index.js
```

but not totally sure...

File: CHANGELOG.md (hunk 15 above)
Comment (ryan-roemer) [inline CHANGELOG.md]: Put Unreleased here. Leave `package.json:version` unchanged.

**Comment by mscottx88** [comment]
@ryan-roemer I reverted the version bump and changed from `normalize` to `resolve`.

File: package.json (hunk 13 above)
Comment (mscottx88) [inline package.json]: Cool, that's what I ended up with too!

File: CHANGELOG.md (hunk 15 above)
Comment (mscottx88) [inline CHANGELOG.md]: Cool, I will correct this one!

**Comment by mscottx88** [comment]
Woohoo!  I figured out how to get the screen to update without the `forced` option in showLayouts.  All I needed was a lil bit of inversion of control.  Turns out the BaseLineGraph needs to request the data, so when changing the zoom level, the higher classes provide a mapper function.

**Comment by jasonwilson** [comment]
The units seem to be offset from whole intervals. Can we make these fixed?
screen shot 2017-07-25 at 4 39 43 pm

**Comment by mscottx88** [comment]
Do you mean, keep them as seconds until everything is evenly divisible by minutes, and so on?

**Comment by mscottx88** [comment]
It turns out

**Comment by mscottx88** [comment]
It turns out that 7.3m = 438s, both of which take up the same space on the screen 🤣

**Comment by jasonwilson** [comment]
I mean so the units are fixed, whole number increments. For example: 1m 5m 10m 15m 20m 25m

**Comment by mscottx88** [comment]
image

**Comment by mscottx88** [comment]
I've got the new #67 ready to go, in a separate branch, but it relies on this one.

**Comment by mscottx88** [comment]
@jasonwilson @aisapatino Hi there, I was just following up on this PR.  I have a few more branches/features I'd like to merge after this one.

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,8 @@
 # Change log
 
+## [Unreleased] - 2017-07-24
+- **Added:** Longer history for graphs [\#64]
```
Comment (jasonwilson) [inline CHANGELOG.md]: Need to include the PR reference at the bottom of the file.

File: lib/providers/metrics-provider.js [hunk 17]
```diff
@@ -1,25 +1,500 @@
 "use strict";
 
+/**
+ * This module provides metric data for various data points.
```
Comment (jasonwilson) [inline lib/providers/metrics-provider.js]: Lets put this writeup in a separate markdown file and abbreviate the explanation here.

File: lib/views/base-line-graph.js [hunk 18]
```diff
@@ -71,6 +77,38 @@ BaseLineGraph.prototype.update = function (values) {
   this.node.setData(_.values(this.series));
 };
 
+BaseLineGraph.prototype.changeMetricData = function (mapper) {
```
Comment (jasonwilson) [inline lib/views/base-line-graph.js]: Anyway to have blessed fix the increments?

File: test/lib/dashboard-agent.spec.js [hunk 19]
```diff
@@ -64,14 +66,31 @@ describe("dashboard-agent", function () {
...
+        expect(metrics.mem.systemTotal).to.equal(20);
+        expect(metrics.mem.rss).to.equal(30);
+        expect(metrics.mem.heapTotal).to.equal(40);
+        expect(metrics.mem.heapUsed).to.equal(50);
+        expect(metrics.cpu.utilization).to.equal(60);
       };
 
+      sandbox.stub(process, "memoryUsage", function () {
+        return {
+          systemTotal: 20,
```
Comment (jasonwilson) [inline test/lib/dashboard-agent.spec.js]: 👍

**Comment by mscottx88** [comment]
@jasonwilson Corrected those, thanks for the recommendation on the separate MD.  I was able to include some visuals which I really wanted to do anyway 👍

**Comment by jasonwilson** [comment]
🚀
